src/test/java/bio/terra/app/usermetrics/BardClientTest.java

```diff
@@ -0,0 +1,169 @@
+package bio.terra.app.usermetrics;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.CALLS_REAL_METHODS;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.withSettings;
+
+import bio.terra.app.configuration.UserMetricsConfiguration;
+import bio.terra.common.category.Unit;
+import bio.terra.common.iam.AuthenticatedUserRequest;
+import java.util.Map;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.SpyBean;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.context.junit4.SpringRunner;
+import org.springframework.web.client.RestTemplate;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(properties = {"datarepo.testWithEmbeddedDatabase=false"})
+@AutoConfigureMockMvc
+@ActiveProfiles({"google", "unittest"})
+@Category(Unit.class)
+@SuppressWarnings("rawtypes")
+public class BardClientTest {
+
+  @SpyBean UserMetricsConfiguration userMetricsConfiguration;
+
+  private AuthenticatedUserRequest user1;
+
+  private AuthenticatedUserRequest user2;
+
+  private String SYNC_PATH;
+
+  private String API_PATH;
+
+  @Before
+  public void setup() {
+    SYNC_PATH = userMetricsConfiguration.getBardBasePath() + "/api/syncProfile";
+    API_PATH = userMetricsConfiguration.getBardBasePath() + "/api/event";
+    user1 =
+        AuthenticatedUserRequest.builder()
+            .setSubjectId("Bob")
+            .setEmail("bob")
+            .setToken("bob token")
+            .build();
+
+    user2 =
+        AuthenticatedUserRequest.builder()
+            .setSubjectId("alice")
+            .setEmail("alice")
+            .setToken("alice token")
+            .build();
+  }
+
+  @Test
+  public void testBardClientLogEvent() {
+    BardClient bardClient =
+        mock(
```

Comment:
It's really hard to test a class well if you need to mock parts of it to test it. If you split the business logic out, or moved the lower level API calls to a different service, it would be easier to write a test and maybe better organized.

Also, is there a benefit to using `defaultAnswer(CALLS_REAL_METHODS)` over `spy()`? I usually write tests like this using `spy()`.

Reply:
> For the spy/mock discussion, I like controlling the constructor arguments. Mock gives me the ability to do that. I'm happy to use @SpyBean if the team requires that approach, although I'm even less of a fan of auto wiring the class under test.

I think `@SpyBean` is something else. This is what I had in mind:
```
    BardClient bardClient = spy(new BardClient(userMetricsConfiguration))
```
I think that does the same thing that your code does.

> I agree with your concerns regarding mock/spy to test, however the AC indicated "Unit only testing (via mocks)"

Splitting the service layer code out from the business logic would allow this. But it would mean adding a new service class. Personally I think the extra class is worth getting rid of a spy or mixed mock/live object test. For example: https://github.com/DataBiosphere/terra-data-catalog/pull/82

> I was debating the addition of [WireMock](https://wiremock.org/). Have you used it before for similar cases?

I've never used it, although I have written and used something like it in the past. IMO I'm skeptical of adding another library until I'm convinced that there's nothing we're already using that could do it.

Reply:
I've updated to use the spy.